### ISSUE TITLE
Make metric functions more generic

### ISSUE THREAD

**Issue by ealt** [Astera-org/simplexity#45]
Currently Metrics are assumed to be an mean value of some function computed over logits and labels, but there are some metrics that should be taken over individual token positions rather than averaged over the sequence for example, or are functions of other values like a distribution over generative process outputs
